### PR TITLE
MockitoInlineMockMaker provides mocking on final classes and methods

### LINKED ISSUE #735: Add support for mocking final classes/methods using ByteBuddy
Mockito 2 has experimental support for mocking final classes/methods (https://github.com/mockito/mockito/wiki/What%27s-new-in-Mockito-2#unmockable). Since Spock as of 1.1 also supports ByteBuddy, it should be possible to also support mocking of final classes/methods.

### PR DESCRIPTION
**Pull request by AndreasTu**
The MockitoMockMaker provides the ability to mock
final classes and final methods with the InlineMockMaker of Mockito.

The MockitoMockMaker would fix #735 and #683.

### CODE REVIEW

**Comment by leonard84** [comment]
Just wondering whether the `@Internal` annotations are not a problem for the inline mock maker?

File: spock-core/src/main/java/org/spockframework/mock/runtime/mockito/MockitoInlineMockMaker.java [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+  }
+
+  @Override
+  public Set<MockMakerCapability> getCapabilities() {
+    return CAPABILITIES;
+  }
+
+  @Override
+  public int getPriority() {
+    return 210;
```
Comment (leonard84) [inline spock-core/src/main/java/org/spockframework/mock/runtime/mockito/MockitoInlineMockMaker.java:66]: 🤔 Maybe we should shift the cglib to 400 and this one to 300?

File: spock-core/src/main/java/org/spockframework/mock/runtime/mockito/MockitoInlineMockMakerImpl.java [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+  private final Method spockMockMethod;
+
+  MockitoInlineMockMakerImpl() {
+    MockitoFramework framework = Mockito.framework();
+    this.inlineMockMaker = framework.getPlugins().getInlineMockMaker();
+    try {
+      this.spockMockMethod = ISpockMockObject.class.getMethod("$spock_get");
+    } catch (NoSuchMethodException e) {
+      throw new IllegalStateException(e);
+    }
```
Comment (leonard84) [inline spock-core/src/main/java/org/spockframework/mock/runtime/mockito/MockitoInlineMockMakerImpl.java:56]:
```suggestion
      this.spockMockMethod = ReflectionUtil.getMethodByName(ISpockMockObject.class, "$spock_get");
```

File: spock-specs/src/test/groovy/org/spockframework/docs/interaction/MockMakerDocSpec.groovy [hunk 3]
```diff
@@ -33,4 +35,36 @@ class MockMakerDocSpec extends Specification {
...
+  }
+
+  @SuppressWarnings('UnnecessaryQualifiedReference')
+  def "mockito-inline additional settings serializable"() {
+    given:
+// tag::mock-serializable[]
+    Subscriber subscriber = Mock(mockMaker: MockMakers.mockitoInline {
+      mockito { org.mockito.MockSettings settings ->
+        settings.serializable()
+      }
```
Comment (leonard84) [inline spock-specs/src/test/groovy/org/spockframework/docs/interaction/MockMakerDocSpec.groovy:59]: ❓ Why did you add this indirection?

File: spock-specs/src/test/groovy/org/spockframework/mock/runtime/ByteBuddyMockFactoryConcurrentSpec.groovy [hunk 4]
```diff
@@ -39,7 +40,7 @@ class ByteBuddyMockFactoryConcurrentSpec extends Specification {
   }
 
   // Just to be save to abort, normally the tests run in 2 secs.
-  @Timeout(40)
+  @Timeout(120)
```
Comment (leonard84) [inline spock-specs/src/test/groovy/org/spockframework/mock/runtime/ByteBuddyMockFactoryConcurrentSpec.groovy:43]: ❓ Why this 3x increase?
40s is already really long compared to all the other tests, I'd like to keep running Spock tests as fast as reasonably possible.

Do we need to separate out these probabilistic tests to some scheduled ci jobs?

File: spock-specs/src/test/groovy/org/spockframework/mock/runtime/ByteBuddyMockMakerSpec.groovy [hunk 5]
```diff
@@ -77,4 +77,72 @@ class ByteBuddyMockMakerSpec extends Specification {
     def ex = thrown(CannotCreateMockException)
     ex.message == "Cannot create mock for class java.lang.StringBuilder. byte-buddy: Cannot mock final classes."
   }
+
+  def "Mock class"() {
+    when:
+    DataClass s = Mock(mockMaker: MockMakers.byteBuddy)
+    then:
+    !s.boolField
+    s.stringField == null
```
Comment (leonard84) [inline spock-specs/src/test/groovy/org/spockframework/mock/runtime/ByteBuddyMockMakerSpec.groovy:86]: ❔ Isn't it a bit confusing calling it field when we are using the getters?
It's also conflating the tests a bit with property handling. I think we should use `bool()` and `string()` instead.

File: spock-specs/src/test/groovy/org/spockframework/mock/runtime/mockito/MockitoInlineMockMakerSpec.groovy [hunk 6]
```diff
@@ -0,0 +1,436 @@
...
+  def "Verify simple ID and IMockMakerSettings"() {
+    expect:
+    mockitoInline.mockMakerId.toString() == ID
+    mockitoInline.toString() == ID
+  }
+
+  def "Verify ID and IMockMakerSettings with Mockito settings"() {
+    when:
+    def set = mockitoInline {}
+    then:
```
Comment (leonard84) [inline spock-specs/src/test/groovy/org/spockframework/mock/runtime/mockito/MockitoInlineMockMakerSpec.groovy:49]: :x: Add a newline before the block labels when starting a new block, to increase readability.

File: spock-specs/src/test/groovy/org/spockframework/mock/runtime/mockito/MockitoInlineMockMakerSpec.groovy [hunk 7]
```diff
@@ -0,0 +1,436 @@
...
+    def ex = thrown(CannotCreateMockException)
+    ex.message == "Cannot create mock for class java.lang.StringBuilder. mockito-inline: Cannot mock final classes with additional interfaces."
+  }
+
+  def "Additional mockito settings serializable"() {
+    when:
+    Runnable mock = Mock(mockMaker: mockitoInline({
+      mockito { serializable() }
+    }))
```
Comment (leonard84) [inline spock-specs/src/test/groovy/org/spockframework/mock/runtime/mockito/MockitoInlineMockMakerSpec.groovy:189]:
```suggestion
    Runnable mock = Mock(mockMaker: mockitoInline {
      mockito { serializable() }
    })
```

**Review by leonard84 — CHANGES_REQUESTED** [review]
Looks great, just a few remarks

File: spock-specs/src/test/groovy/org/spockframework/docs/interaction/MockMakerDocSpec.groovy (hunk 3 above)
Comment (AndreasTu) [reply spock-specs/src/test/groovy/org/spockframework/docs/interaction/MockMakerDocSpec.groovy:59]: To be able to add mockSettings for the Spock side in the future.
E.g. that we could configure, how Spock would use/create the Mockito-Mock.
If I would use the `org.mockito.MockSettings` directly, we could not add any settings to that interface.

I was also wondering, if I should rename the mock maker from `mockito-inline` to `mockito`, because that I am using the `inline` part, would be an implementation details, which could for example be configured with the mock settings.

File: spock-specs/src/test/groovy/org/spockframework/mock/runtime/ByteBuddyMockFactoryConcurrentSpec.groovy (hunk 4 above)
Comment (AndreasTu) [reply spock-specs/src/test/groovy/org/spockframework/mock/runtime/ByteBuddyMockFactoryConcurrentSpec.groovy:43]: The tests are flaky under load on my machine, so I increased the timeout, but this is not a real fix.
Maybe we would need to make them `@Isolated` or separate them as you proposed.

If I run them without any other tests, they are fast, but with congestion, they tend to break the timeout.

File: spock-specs/src/test/groovy/org/spockframework/docs/interaction/MockMakerDocSpec.groovy (hunk 3 above)
Comment (leonard84) [reply spock-specs/src/test/groovy/org/spockframework/docs/interaction/MockMakerDocSpec.groovy:59]: 🤔 I would go with YAGNI, unless you already have a concrete plan to add something more here. If we go with your suggestion of renaming it, which is ok with me, then it looks even weirder
```
MockMakers.mockito {
  mockito {
    settings.serializable()
  }
}
```

If we really need to add something in the future, we could use a differently named method to offer the more complex config.

Comment (AndreasTu) [reply spock-specs/src/test/groovy/org/spockframework/docs/interaction/MockMakerDocSpec.groovy]: >If we really need to add something in the future, we could use a differently named method to offer the more complex config.

That is a good point.
If have removed the settings indirection and renamed `mockito-inline` to `mockito`

**Comment by leonard84** [comment]
Thanks for hanging in there @AndreasTu
